tests/test_source.py

```diff
@@ -127,6 +129,15 @@ def test_get_package_hashes_server(self):
              'sha256': '795a1bdddc832289ab958dc9bd9a1c46b849011cbc81fd89d6fe144efc7aae69'}
         ]
 
+        with ZipFile(os.path.join(os.getcwd(), 'tests/data/tensorflow_serving_api-1.13.0-py2.py3-none-any.whl'), 'r') as zip:
+            with tempfile.TemporaryDirectory() as tmpdir:
+                artifact_name = 'tensorflow_serving_api-1.13.0-py2.py3-none-any'
+                zip.extractall(os.path.join(tmpdir, artifact_name))
+                with open(os.path.join(os.getcwd(), 'tests/data/tensorflow_serving_api-1.13.0-py2.py3-none-any.json')) as json_file:
+                    result = json.load(json_file)
+                    source = Source.from_dict(pypi_index)
+                    assert source._construct_contents(os.path.join(tmpdir, artifact_name))== result
```

Comment:
As we not downloading .whl file from an index, we are not using```source._get_package_hashes()``` method because it will trigger a download instead we are just sending the path of our binary .whl file which is in test/data/ directory.

Reply:
Tests like this are usually done using "mocking". Use [flexmock](https://pypi.org/project/flexmock/) library, we use it in other components in the project (see for example adviser test suite).

First, install flexmock:

```
pipenv install --dev flexmock
```

and the test could look something like this (not tested, just to show the idea):

```python
import flexmock

flexmock(Source).should_receive("_get_contents").with_args("tensorflow", "1.13.0").and_return(result)
```

Reply:
```source = Source.from_dict(pypi_index)``` part is skipped. Instead just ```Source``` class is used as a parameter to flexmock initialization. ```_construct_contents``` method is used with an argument being a path to the extracted wheel folder.